server/src/main/java/org/elasticsearch/gateway/GatewayService.java

```diff
@@ -141,83 +144,162 @@ public void clusterChanged(final ClusterChangedEvent event) {
+     *     <li>Or schedule a recovery with a delay of {@code recoverAfterTime}
+     * </ul>
+     */
+    class PendingStateRecovery {
+        private final long expectedTerm;
+        @Nullable
+        private volatile Scheduler.ScheduledCancellable scheduledRecovery;
 
-                    @Override
-                    protected void doRun() {
-                        if (recoveryInProgress.compareAndSet(false, true)) {
-                            logger.info("recover_after_time [{}] elapsed. performing state recovery...", recoverAfterTime);
-                            runRecovery();
-                        }
-                    }
-                }, recoverAfterTime, threadPool.generic());
+        PendingStateRecovery(long expectedTerm) {
+            this.expectedTerm = expectedTerm;
+        }
+
+        void onDataNodeSize(int currentDataNodeSize) {
+            if (recoverAfterDataNodes != -1 && currentDataNodeSize < recoverAfterDataNodes) {
+                logger.debug(
+                    "not recovering from gateway, nodes_size (data) [{}] < recover_after_data_nodes [{}]",
+                    currentDataNodeSize,
+                    recoverAfterDataNodes
+                );
+                cancelScheduledRecovery();
+            } else {
+                maybePerformOrScheduleRecovery(currentDataNodeSize);
             }
-        } else {
-            if (recoveryInProgress.compareAndSet(false, true)) {
-                try {
-                    logger.debug("performing state recovery...");
-                    runRecovery();
-                } catch (Exception e) {
-                    logger.warn("state recovery failed", e);
-                    resetRecoveredFlags();
+        }
+
+        void maybePerformOrScheduleRecovery(int currentDataNodeSize) {
+            if (expectedDataNodes != -1 && expectedDataNodes <= currentDataNodeSize) {
+                logger.debug(
+                    "performing state recovery of term [{}], expected data nodes [{}] is reached",
+                    expectedTerm,
+                    expectedDataNodes
+                );
+                runRecoveryImmediately();
+            } else if (recoverAfterTime == null) {
+                logger.debug("performing state recovery of term [{}], no delay time is configured", expectedTerm);
+                runRecoveryImmediately();
+            } else {
+                if (scheduledRecovery == null) {
+                    logger.info(
+                        "delaying initial state recovery for [{}] of term [{}]. expecting [{}] data nodes, but only have [{}]",
+                        recoverAfterTime,
+                        expectedTerm,
+                        expectedDataNodes,
+                        currentDataNodeSize
+                    );
+                    scheduledRecovery = threadPool.schedule(getScheduleTask(), recoverAfterTime, threadPool.generic());
```

Comment:
@DaveCTurner There are still some complexities with this block of code and other related areas.
1. This class has no state to remember that it has attempt to recover immediately. Therefore if the cluster can recover immediately, it can potentially submit multiple cluster update tasks. Is it a problem?
2. For the scheduled case, we can mostly avoid multiple schedules by checking whether `scheduledRecovery` is `null`. There can still be edge cases where we can schedule more than once due to racing between checking `scheduledRecovery` and reset it back to `null`. If submitting multiple update tasks isn't an issue, we may also chose to not check it all and just always schedule?
3. Because we need to reset `scheduledRecovery` back to null in the scheduled runnable, it needs to be made `volatile` as well.

Do we need to address the 1st point? If so, it seems we need another state variable for it. I forgot to mention it during the sync but this was one of the original complexity. Also because the `ClusterStateUpdateTask` may not run inside `execute` due to dataNodeSize dropping again, the state needs to be reset from within the task which brings back the need of passing a "runAfter" into the task. To simplify things, I think we don't want to check dataNodeSize again inside the task. It's an edge case anyway and dropping it makes things simpler. But we will still need some other state management if we want to address the 1st point. What do you think?

Reply:
1. Good point. I think we should only ever submit one cluster state update task per term, so we ought to track this with a flag within the per-term state.

2. I would not expect any races here, or rather I think if we keep track of whether we've submitted the cluster state update task then that solves those races.

3. Good point, although since we only do that when actually submitting the task again I think the solution is to make this submission a once-only thing.

Reply:
Pushed 25bbebaa76e715918f613c3666827a9248f0f530 to add a new state variable (`AtomicBoolean`) solved multiple issues that I had. Thanks!

Please let me know if the main code looks good to you.  I'll proceed to add some more tests if you are happy with the main code changes. Please also let me know if you have any ideas for what kinda tests we might need. Thanks!

Reply:
I have now added multiple tests to cover different scenarios in bf8f21ac0d2145299a94d003eb9909732eb885f2
Now the whole thing looks ready to me.